### ISSUE TITLE
Kamigawa tap lands don't work correctly when changing control.

### ISSUE THREAD

**Issue by Grim-the-Reaper** [Card-Forge/forge#984]
When you activate a Kamigawa tap land's second ability, and then an opponent takes control of it, it should untap during their untap step. If you later steal it back, it should not untap during yours. The 'your' refers to the controller of the mana ability, not the controller of the land, so the effect needs to keep track of who activated the ability that generates this effect (and the land could have multiple such effects active for different players, since they could tap it for that ability too, which would only affect their untap step).

This should probably be looked at for all 'doesn't untap' effects, they should all refer to a specific player. (who could be, but not necessarily still is, that permanent's controller in the untap step.) (So this a major issue which affects a lot of cards.)

In other words, this should become 'This permanent doesn't untap during [Player X]'s next untap step.'
with [PlayerX] being a variable.

Another thing that should be taken into account is that this effect should end when that player's next untap step happens, even if they don't control the land.

(I also saw in the code that the current text used for this is using 'this card' instead of 'this permanent', which is technically incorrect, since that doesn't cover tokens. I know this isn't that important code-wise, but I think using the correct wording can only be beneficial, for clarity's sake.)

**Comment by Hanmac** [comment]
the reason is that the hidden keyword "doesn't untap" should be changed into a static effect created by the Activator of the Ability

**Comment by Grim-the-Reaper** [comment]
If you go that road, couldn't you just make the effect invisible to the UI, so nothing gets clogged up?

Note that the one who controlled the activated / triggered / spell ability that generated such an effect is not always the one affected by it, though, so knowing who created the effect is not enough, you need to actually track the affected player too.

Abilities that read 'That permanent doesn't untap during its controller's next untap step.' are an exception.
In that case, it will trigger for whoever controls that creature at the time of untap, and does not predetermine the player at the time of effect creation, like these Kamigawa lands do.

**Comment by tool4ever** [comment]
Old branch in case someone wants to pick it up:
https://github.com/Card-Forge/forge/tree/untapRework

**Comment by tool4ever** [comment]
The new one tracks what remains of this issue
